nntrainer/compiler/compiler.h

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file compiler.h
+ * @date 01 April 2021
+ * @brief NNTrainer compiler that reads to generate optimized graph
+ * @details
+ * Graph is convertible either to iostream, representation, executable by
+ * appropriate compiler and interpreter
+ * For example, if istream would be from a a.tflite file,
+ *
+ * GraphRepresentaion g;
+ * GraphCompiler * compiler = new NNTrainerCPUCompiler;
+ *
+ * ExecutableGraph eg = compiler->compile(g);
+ *
+ *
+ *    +-------+--+--------+
+ *    |GraphRepresentation|
+ *    +-------+-----------+
+ *            |  ^
+ *  compile() |  |
+ *            |  |
+ *         (Compiler)
+ *            |  |
+ *            |  | decompile()
+ *            v  |
+ *      +--------+------+
+ *      |ExecutableGraph|
+ *      +---------------+
+ *
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+#ifndef __COMPILER_H__
+#define __COMPILER_H__
+
+#include <memory>
+
+#include <network_graph.h>
+
+namespace nntrainer {
+
+using GraphRepresentation = NetworkGraph;
+using ExecutableGraph = NetworkGraph;
+
+/**
+ * @brief Pure virtual class for the Graph Compiler
+ *
+ */
+class GraphCompiler {
+public:
+  virtual ~GraphCompiler() {}
+  /**
+   * @brief serialize graph to a file stream
+   *
+   * @param representation graph representation
+   * @param file ifstream to serialize graph
+   */
+  virtual std::shared_ptr<ExecutableGraph>
+  compile(std::shared_ptr<const GraphRepresentation> representation) = 0;
```

Comment:
add a todo to take a delegate parameter in this function or in constructor?

Reply:
I am thinking to delegate through a interface. This might be done outside of the class :)

Reply:
delegate contains two parts - backend `B` (nntrainer, tflite, nnapi etc) and hardware `H` (cpu, gpu, etc). Do you plan to create  `B*H` interfaces?

Reply:
I was thinking B*H.  (nntrainerCPU, nntrainerGPU), as they would have different implementation. If considerable amount of codes can be shared, it might be better to have a delegate inside the constructor.

Reply:
Starting with either way works. FYI: with `B*H` interfaces, `delegate` will be resolved in the model, when creating the compiler object.

Reply:
Done

Reply:
I wonder that if you think nntrainerCPU, nntrainerGPU.. then how do you handle CPU+GPU Case?? or heterogeneous case??

Reply:
We discussed offline that this is very complicated problem and shouldn't be fixated without comprehended survey. This will remain unfixed and codes will consider that as well.